torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -8820,4 +8820,100 @@ def celu_hook(module, _input, output):
             self.assertEqual(ddp_grads[0], local_model.fc.weight.grad)
             self.assertEqual(ddp_grads[1], local_model.fc.bias.grad)
 
+        def _test_hook_pickling(self, hook, hook_state):
+            learning_rate = 0.01
+            chkpt_file = tempfile.gettempdir() + "/checkpoint.pt"
+            rank = self.rank
+
+            input = torch.randn(1, 1, device=rank)
+            net = torch.nn.Linear(1, 5).to(rank)
+            ddp_model = DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            dummy_ddp_model = DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            optimizer = torch.optim.SGD(ddp_model.parameters(), lr=learning_rate)
+            ddp_model.register_comm_hook(hook_state, hook)
+
+            for _ in range(10):
+                optimizer.zero_grad()
+                loss = ddp_model(input).sum()
+                loss.backward()
+                optimizer.step()
+
+            state = {
+                'state_dict': ddp_model.state_dict(),
+                'comm_hook': hook,
+                'comm_hook_state': hook_state
+            }
+
+            if rank == 0:
+                torch.save(state, chkpt_file)
+
+            dist.barrier()
+            map_location = {'cuda:%d' % 0: 'cuda:%d' % rank}
+            with self.assertLogs() as captured:
+                checkpoint = torch.load(chkpt_file, map_location=map_location)
+
+            # Check that the logger has only one entry
+            self.assertEqual(len(captured.records), 1)
+            # Check that the logger has an expected entry
+            self.assertEqual(
+                captured.records[0].getMessage(),
+                'NOTE: Current process group is set to default.'
+            )
+
+            dummy_ddp_model.load_state_dict(checkpoint['state_dict'])
```

Comment:
It might be valuable to register the loaded hook + state here and ensure that works without any issues.

Further, the test could be more comprehensive if we run some iterations on this reloaded version and ensure we have parity with before checkpoint (you can check the gradients, and make sure everything starts off / is seeded the same)?

Reply:
Thanks for suggestion! I've refactored unit test -- i.e. adjusted model initialization and training procedure, added extra training after a model was reloaded, added check on gradients as suggested.